components/camera/rtsp/rtsp.go

```diff
@@ -68,170 +60,80 @@ func (conf *Config) Validate(path string) ([]string, error) {
-	}
+func (r *rtsp) Close(_ context.Context) error {
+	r.cancelFunc()
+	r.activeBackgroundWorkers.Wait()
+
 	return nil
 }
 
-// clientReconnectBackgroundWorker checks every 5 sec to see if the client is connected to the server, and reconnects if not.
-func (rc *rtspCamera) clientReconnectBackgroundWorker() {
-	rc.activeBackgroundWorkers.Add(1)
-	goutils.ManagedGo(func() {
+type imgError struct {
+	image.Image
+	error
+}
+
+// Reading from an RTSP stream as fast as possible prevents the write queue from filling up which results in the stream
+// dropping the session and EOF errors. This should only be a problem for recorded streams or streams that feeds us
+// frames faster than we can process them.
+func (r *rtsp) startStreaming(ctx context.Context, addr string) (chan imgError, error) {
+	imgCh := make(chan imgError)
+
+	dec, err := NewDecoder(ctx, addr)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot start rtsp stream")
+	}
+
+	r.activeBackgroundWorkers.Add(1)
+	utils.ManagedGo(func() {
+		defer dec.Close()
 		for {
-			if ok := goutils.SelectContextOrWait(rc.cancelCtx, 5*time.Second); ok {
-				// use an OPTIONS request to see if the server is still responding to requests
-				res, err := rc.client.Options(rc.u)
-				badState := false
-				if err != nil && (errors.Is(err, liberrors.ErrClientTerminated{}) ||
-					errors.Is(err, io.EOF) ||
-					errors.Is(err, syscall.EPIPE) ||
-					errors.Is(err, syscall.ECONNREFUSED)) {
-					rc.logger.Warnw("The rtsp client encountered an error, trying to reconnect", "url", rc.u, "error", err)
-					badState = true
-				} else if res != nil && res.StatusCode != base.StatusOK {
-					rc.logger.Warnw("The rtsp server responded with non-OK status", "url", rc.u, "status code", res.StatusCode)
-					badState = true
-				}
-				if badState {
-					if err = rc.reconnectClient(); err != nil {
-						rc.logger.Warnw("cannot reconnect to rtsp server", "error", err)
-					} else {
-						rc.logger.Infow("reconnected to rtsp server", "url", rc.u)
-					}
-				}
-			} else {
+			select {
+			case <-r.cancelCtx.Done():
 				return
+			default:
 			}
-		}
-	}, rc.activeBackgroundWorkers.Done)
-}
 
-// reconnectClient reconnects the RTSP client to the streaming server by closing the old one and starting a new one.
-func (rc *rtspCamera) reconnectClient() (err error) {
-	if rc == nil {
-		return errors.New("rtspCamera is nil")
-	}
-	if rc.client != nil {
-		err := rc.client.Close()
-		if err != nil {
-			rc.logger.Debugw("error while closing rtsp client:", "error", err)
-		}
-	}
-	// replace the client with a new one, but close it if setup is not successful
-	client := &gortsplib.Client{}
-	rc.client = client
-	var clientSuccessful bool
-	defer func() {
-		if !clientSuccessful {
-			if errClose := rc.client.Close(); errClose != nil {
-				err = multierr.Combine(err, errClose)
+			img, err := dec.Decode(r.cancelCtx)
+			select {
+			case <-r.cancelCtx.Done():
+				return
+			case imgCh <- imgError{img, err}:
```

Comment:
[q] If we get `EOF` or other errors here, would we need to reconnect or reestablish the encoder?

Reply:
The decoder you mean? Yes. If we start getting errors we should probably reconnect to the RTSP address. Looking at the output from the docker RTSP server, it looks like when EOF errors happen the server-side session is closed and therefore needs to be-reestablished. Sending a rebuild error may work but could be slower than doing it in-house.